src/components/multi-select/multi-select.jsx

```diff
@@ -0,0 +1,316 @@
+      .filter(option => !value.some((val) => {
+        const optionID = props.optionIDGetter(option);
+
+        if (typeof optionID === 'string') {
+          return val.toLowerCase() === optionID.toLowerCase();
+        }
+
+        return val === optionID;
+      }));
+  }
+
+  function onAutocompleteBlur() {
+    if (!autocompleteRef.current) return;
+
+    autocompleteRef.current.setCustomValidity('');
+    if (!autocompleteRef.current.validity.valid) {
+      setValidationMessage(autocompleteRef.current.validationMessage);
+    } else {
+      setValidationMessage(props.validationMessage || '');
+    }
+  }
+
+  function onAutocompleteChange(event) {
+    setExpanded(true);
+    setAutocompleteValue(event.target.value);
+  }
+
+  function onDropdownClose() {
+    setExpanded(false);
+    setAutocompleteValue('');
+  }
+
+  function onOptionRemove(event) {
+    const newValue = value.filter((option, index) => (
+      valueRefs[index].current !== event.target
+    ));
+    setValue(newValue);
+  }
+
+  function onOptionsClear(event) {
+    event.preventDefault();
+
+    if (!props.readOnly) {
+      setValue([]);
+      setExpanded(false);
+      autocompleteRef.current.focus();
+    }
+  }
+
+  function onOptionSelect(event) {
+    const selectedOptionIndex = visibleOptionsRefs.findIndex(optionRef => optionRef === event.target);
+    const selectedOption = visibleOptions[selectedOptionIndex];
+    setExpanded(false);
+    const newValue = [...value, props.optionIDGetter(selectedOption)]
+      .sort((a, b) => props.optionIDGetter(a) - props.optionIDGetter(b));
+    setValue(newValue);
+    setAutocompleteValue('');
+    autocompleteRef.current.focus();
+  }
+
+  function onClick(event) {
+    if (event.defaultPrevented) return;
+    if (props.readOnly) return;
+
+    setExpanded(true);
+  }
+
+  function onKeyDown(event) {
+    switch (event.key) {
+      case 'Escape':
+        setExpanded(false);
+        break;
+      default:
+    }
+  }
+
+  useDropdownClose(wrapperRef, expanded, onDropdownClose);
+
+  useEffect(() => {
+    setValue(props.value);
+  }, [props.value.map(val => JSON.stringify(val)).join(',')]);
+
+  useEffect(() => {
+    props.onChange({ target: selfRef.current });
+  }, [value.map(val => JSON.stringify(val)).join(',')]);
+
+  useImperativeHandle(selfRef, () => ({
+    get dirty() {
+      return props.value.map(val => JSON.stringify(val)).join(',') !== this.value.map(val => JSON.stringify(val)).join(',');
+    },
+    id: props.id,
+    name: props.name,
+    get value() {
+      return value;
+    },
+  }));
+
+  return (
+    <div className={classNames.container} id={props.id} ref={wrapperRef}>
+      <FormControl
+        label={props.label}
+        labelId={ids.label}
+        id={ids.textbox}
+        error={validationMessage}
+        onKeyDown={onKeyDown}
+      >
+        <div role="presentation" className={classNames.formControlChildrenContainer} onClick={onClick}>
```

Comment:
If the `<FormControlIcon />` is absolutely positioned, should we consider removing this wrapping div in favor of customizing `<FormControl />` with a better class names API and perhaps an `onClick`?

Reply:
That's an interesting proposition. I think we should keep this in mind for the future, and if we see this kind of use-case keep coming back up, then consider extending `FormControl`. As it is, I don't think this is a bad thing to do here considering the complexity of our use-case, and keeping `FormControl` as simple as possible is a win in my book.